crates/aptos-telemetry-service/src/peer_location.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_infallible::RwLock;
+use aptos_types::PeerId;
+use gcp_bigquery_client::{model::query_request::QueryRequest, Client as BigQueryClient};
+use std::{collections::HashMap, str::FromStr, sync::Arc, time::Duration};
+
+const ANALYTICS_PROJECT_ID: &str = "analytics-test-345723";
+
+#[derive(Clone, Debug)]
+pub struct PeerLocation {
+    pub peer_id: PeerId,
+    pub country: Option<String>,
+    pub region: Option<String>,
+}
+
+pub struct PeerLocationUpdater {
+    client: BigQueryClient,
+    peer_locations: Arc<RwLock<HashMap<PeerId, PeerLocation>>>,
+}
+
+impl PeerLocationUpdater {
+    pub fn new(
+        client: BigQueryClient,
+        peer_locations: Arc<RwLock<HashMap<PeerId, PeerLocation>>>,
+    ) -> Self {
+        Self {
+            client,
+            peer_locations,
+        }
+    }
+
+    pub fn run(self) -> anyhow::Result<()> {
+        tokio::spawn(async move {
+            loop {
+                let locations = query_peer_locations(&self.client).await.unwrap();
+                {
+                    let mut peer_locations = self.peer_locations.write();
+                    *peer_locations = locations;
+                }
+                tokio::time::sleep(Duration::from_secs(3600)).await; // 1 hour
+            }
+        });
+        Ok(())
+    }
+}
+
+pub async fn query_peer_locations(
+    client: &BigQueryClient,
+) -> anyhow::Result<HashMap<PeerId, PeerLocation>> {
+    let mut res = client.job().query(ANALYTICS_PROJECT_ID, QueryRequest::new("
+            SELECT
+                sq.peer_id,
+                sq.country,
+                sq.region
+            FROM (
+                SELECT
+                tm.peer_id,
+                tm.epoch,
+                ROW_NUMBER() OVER (PARTITION BY tm.peer_id ORDER BY tm.epoch DESC) AS row_number,
+                tm.country,
+                tm.region
+                FROM
+                `node-telemetry.aptos_node_telemetry.custom_events_mainnet_telemetry_rollup_metrics` tm) sq
+            WHERE
+                sq.row_number = 1
+            LIMIT
+                1000
+        "))
+        .await?;
+
+    let mut map = HashMap::new();
+    while res.next_row() {
```

Comment:
can we count if a query failed into prometheus? Also it would be good to timeout the query if it takes very long.